### ISSUE TITLE
Arithmetic on enumerators yields different type

### ISSUE THREAD

**Issue by BryantLam** [chapel-lang/chapel#8552]
### Summary of Problem
Is there a purpose to setting the internal representation of an `enum` if arithmetic yields the default `int(64)`? Also, why does that happen?

### Steps to Reproduce
```chpl
type T_E = uint(8);
enum E {a = 1: T_E, b, c};

var e/*: T_E */ = E.a + E.b + E.c;

writeln("expected 6: uint(8)");
writeln("output ", e, ": ", e.type: string); // 6: int(64)
```

### Configuration Information
`chpl version 1.17.0 pre-release (2f0af5c5db)`

**Comment by bradcray** [comment]
I may be mistaken, but I believe that today, enums' integer values are always treated as though they are `int(64)` values.  Assuming, I'm right about that, the expression `a = 1: T_E` is a lot like saying `var a: int(64) = 1: T_E;`, which is reasonable: a `uint(8)` expression can coerce to an `int(64)`, but assigning an `int(64)` variable a `uint(8)` value doesn't make it one.

I say "_believe_" above because the compiler's treatment of enums has evolved over time and in some respects still isn't what I think we'd like it to be (e.g., search on open issues with [`enum` in their title](https://github.com/chapel-lang/chapel/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+enum+in%3Atitle) ).  So along the way, I feel I've personally lost track of where they stand in some regards.

My leading mental model for `enum`s is that they're like a named, related collection of `param int(64)`s.  I've also personally kicked around the notion of proposing an extension to the current `enum` declaration syntax that would indicate the underlying integral type that an enum should behave like, such as:

```chapel
enum colors: uint(8) = {red, green, blue};
```

This could be a way of getting the behavior that I think you were looking for.  I've never proposed this, as there've always been bigger fish to fry...  But I think it would round out the current enum story well in an "enums are like param integers" interpretation.

Another interpretation that has been kicked around at times was "an enum should behave like the smallest integer width that can store all of their values" — I'm not keen on this primarily because it seems a little fiddly/unstable w.r.t. what the set of enum values are, and also because it seems inconsistent with Chapel using 64-bit types for most things by default.

**Comment by BryantLam** [comment]
> I believe that today, enums' integer values are always treated as though they are int(64) values.

From #5417, I would have believed `a = 1: T_E` set the enum representation. Your explanation makes more sense.

**Comment by bradcray** [comment]
Thanks for the reference back to #5417.  I've been scrambling to try and understand better what's happening here, and think I've mostly worked it out / reminded myself of where we're at:

* First of all, if a user were to have asked me prior to this issue "Is dispatch on enums implemented correctly today?" I would've answered "No, to our regret, I don't think it is", and I think this issue and my notes below reinforce that belief.

* I think that Ben's incorrect when he states in issue #5417 that casting the integer initializer for an enum sets the underlying type of the enum as a whole (or maybe it was true then but no longer is).  At first, I thought he was right because I created a version starting from your code, looked at the generated code, and found indications that all of the enum symbols were stored as uint(8)s there.  However, even if I cast to other things like uint(32) (as Ben does in #5417), I still see them getting stored as uint(8)s in the generated code.  So I think that Ben's claim was either incorrect or is at least no longer true and have followed up on that issue to see if he has an explanation.

* I was also surprised to see that the idea I mention above of explicitly associating an integer type with an enum was mentioned on that issue and am curious whether I mentioned it to Ben or whether great minds think alike.  :)

* If I store an integer value into one of the `enum`s which exceeds a `uint(8)`, the generated code bumps up to a larger integer size when representing the values in the generated code:

  ```chapel
  enum color {red = 257, green, blue};  // uses uint(16) values to store the enum
  enum color {red = -10, green, blue};  // uses int(8) values to store the enum
  enum color {red = -257, green, blue};  // uses int(16) values to store the enum
  ```

  (This may be what I was thinking of above when I mentioned the notion of having enums be like the smallest integer size required to store them).  That said, I think that this only impacts how the enums are represented in the generated code, not how they behave within Chapel (I don't really like that we do this personally, as I'd prefer to have the codegen type be consistent with the semantic type and because I think having the size of your enum automatically change on you depending on what integer values you store in it is more surprising than useful...  particularly if the user had a way to specifically request a particular integer width as proposed here and in issue #5417).

* Interestingly, I found that if I add a variable `uint(8)` to your enum, I get a `uint(8)` result:

  ```chapel
  type T_E = uint(8);
  enum E {a = 1: T_E, b, c};

  var one: uint(8) = 1;
  var e/*: T_E */ = E.a + one;

  writeln("output ", e, ": ", e.type: string); // 2: uint(8)                      
  ```

* However, if I add a `param uint(8)` to your enum, I get an `int(64)` result:

  ```chapel
  type T_E = uint(8);
  enum E {a = 1, b= 2:T_E, c};

  param one: uint(8) = 1;
  var e/*: T_E */ = E.b + 1;

  writeln("output ", e, ": ", e.type: string); // 2: int(64)                      
  ```

* I also find that if I pass enums to an overloaded routine, they prefer the `int(64)` versions regardless of whether `t` below is set to `uint(8)`, `uint(32)` or `uint(64)`:

  ```chapel
  config type t = uint(32);

  enum color { red = 1: t, green, blue};

  proc foo(x: uint(8)) {
    writeln("In uint(8) version");
  }

  proc foo(x: uint(32)) {
    writeln("In uint(32) version");
  }

  proc foo(x: uint(64)) {
    writeln("In uint(64) version");
  }

  proc foo(x: int(64)) {
    writeln("In int(64) version");
  }

  foo(color.red);             // calls int(64) version                            
  foo(color.red+color.green); // calls int(64) version                            
  ```

* Lastly, consider the results of the following combinations:

  ```chapel
  var ui1 = 1: uint(8), ui2 = 1: uint(8);
  param pui1 = 1: uint(8), pui2 = 1: uint(8);
  var i64	= 1;
  param pi64 = 1;

  writeln(ui1.type:string);             // uint(8)                                
  writeln(pui1.type:string);	      // uint(8)                                
  writeln((ui1+ui2).type:string);       // uint(8)                                
  writeln((pui1+pui2).type:string);     // uint(8)                                
  writeln((ui1+1).type:string);         // uint(8) -- param int(64)s defer        
  writeln((pui1+1).type:string);        // uint(8) -- "                           
  writeln((ui1+pi64).type:string);      // uint(8) -- "                           
  writeln((pui1+pi64).type:string);     // uint(8) -- "                           
  writeln((ui1+i64).type:string);       // int(64) -- uint(8)s coerce             
  writeln((pui1+i64).type:string);      // int(64) -- "                           
  ```

The net effect of all of this suggests to me that the compiler is treating enum symbols as `var int(64)` values rather than the `param int(64)` values that I would expect when resolving function overloads.  And that if they were treated more like `param int(64)` values (and had  you not read the misleading information in issue #5417), they'd arguably behave more as you'd expect.

Note that a _variable_ of `enum` type (e.g., `var shade: color;`) should and would continue to be treated as a `var int(64)` and that the previous statement applies only to literal enum symbols (like `color.red`).  By extension, in the proposal that we support explicit type specifiers for enums like `enum color: t { ... }` the idea is that the enumerated symbols would be `param t`s and that a variable of type `color` would be a `var t`.

Tagging @mppf on this issue as typing it up reminds me that I believe he might have worked in a related area of function resolution recently.

**Comment by mppf** [comment]
I have a hunch that extending `isNumericParamDefaultType` to consider enums to default to `int(64)` would solve the ambiguity problem. (perhaps around https://github.com/chapel-lang/chapel/blob/master/compiler/resolution/functionResolution.cpp#L1427).

**Comment by bradcray** [comment]
I was rushing to get all my observations into this issue before timing out at work tonight, still left 20 minutes late, and then on the way home realized that I'd made some mistakes (not including the one that Michael points out above—thanks Michael).  I'd then hoped to get online and edit my response before anyone on the east coast read it, but by the time the girls were in bed, Michael had already commented.  Rather than going back and editing, which might confuse the history, I'll try to correct myself a bit here.

My realization on the way home was that the enum values _are_ acting more like param `int(64)`s than I had characterized in my previous response, which I think is good news.  Specifically, adding the `var uint(8)` to the enum in my first code block and having a result be a `uint(8)` is similar to what `param int(64)`s do in general—defer to alternative type sizes since, as the default, they probably don't represent an intention to indicate a specific size (e.g., it'd be a shame if `myUint8 + 1` favored the `int(64)`.  On reflection, that case was equivalent to the `writeln((ui1+pi64).type:string);` case in my cross-product of cases example, simply with the arguments reversed (which shouldn't affect the outcome).  So that seems encouraging.

Another way to see that they're acting like params is the following test which I should've thought of earlier:

```chapel
enum color {red, green, blue};

proc foo(x: int(64)) {
  writeln("In normal version");
}

proc foo(param x: int(64)) {
  writeln("In param version");
}

foo(color.red);   // calls param version

var hue: color;
foo(hue);         // calls normal version

param shade: color = color.green;
foo(shade);       // calls param version
```

So that's all encouraging in that enums seem to be acting like `int(64)`s (the default integer size) and enum symbols themselves and `param` enum symbols seem to be acting like `param int(64)`s.  So the only case that now seems odd to me is the ambiguous one that Michael noted:

```chapel
enum E {a=1}
param one8 = 1:uint(8);
writeln((E.a+one8).type:string);   // ambiguous?!?
```

Since replacing the `E.a` with a `param int(64)` does not result in a similar ambiguity:

```chapel
param one64 = 1: int(64);
param one8 = 1:uint(8);
writeln((one64+one8).type:string);  // uint(8)
```

I think this suggests "bug" for this case (though I'm pretty surprised that it behaves this way when so much of the rest of enum behavior seems to be matching `param int(64)` so closely).

Returning to the question that opened this issue, I'm inclined to (1) call this issue an "ain't broke / won't fix", (2) edit issue #5417 which started the misunderstanding to correct its apparent mis-statement once Ben confirms, and (3) open a new issue regarding the resolution ambiguity.  Sound OK to you, @BryantLam and @mppf ?

Reflecting a bit more on the variable representation size of enums in codegen...  I'm feeling less opposed to it now than I was a few hours ago, though I feel like I can make arguments in either direction, for or against it.  I may spawn off another issue to try and capture those thoughts even though I'm not objecting as strenuously to the current approach at this moment.

**Comment by bradcray** [comment]
> I may spawn off another issue to try and capture those thoughts even though I'm not objecting as strenuously to the current approach at this moment.

I've done this now, in issue #8561.  As you'll see, by the time I was done with it I felt more against the current approach again, though not out of maliciousness.

**Comment by bradcray** [comment]
A couple of times on this thread, I've expressed concern for our current support for enums, but scanning through the open issues, a few that were contributing to my sense of unease seem to have been resolved yet not marked as such (#5845 and #6208).  In addition, taking a look at #5416, it seems to have a reasonably straightforward fix.  So maybe things aren't as bad as I'd feared.

**Comment by BryantLam** [comment]
> I'm inclined to (1) call this issue an "ain't broke / won't fix", (2) edit issue #5417 which started the misunderstanding to correct its apparent mis-statement once Ben confirms, and (3) open a new issue regarding the resolution ambiguity.

Sounds okay with me. Tag and close as appropriate.

**Comment by bradcray** [comment]
Tying this up, Ben has edited issue #5417 to avoid misleading future readers, I've filed issue #8658 to capture the ambiguity that Michael found above (and have noted that the issue will be moot if we make enums more of a strong type), have labeled this issue "Ain't broke" and am closing it.  Thanks for the discussion and sorry again for the confusion.
